Added Symbolic Link Support

NoMoreFood opened the pull request:
- Created readlink() implementation.
- Separated, corrected lstat() implementation.
- Added symlink file type indicator to returned request data.

NoMoreFood committed:
Added Symbolic Link Support

- Created readlink() implementation.
- Separated, corrected lstat() implementation.
- Added symlink file type indicator to returned request data.

bagajjal commented on contrib/win32/win32compat/fileio.c:
```diff
@@ -749,13 +752,98 @@ fileio_fstat(struct w32_io* pio, struct _stat64 *buf)
 	return _fstat64(fd, buf);
 }
 
+wchar_t * fileio_readlink_internal(wchar_t * wpath)
```
This project follows openBSD standards.

return type, filename should be on different lines.

NoMoreFood replied:
Addressed in updated commit.

bagajjal commented on contrib/win32/win32compat/fileio.c:
```diff
@@ -749,13 +752,98 @@ fileio_fstat(struct w32_io* pio, struct _stat64 *buf)
+		USHORT PrintNameOffset;
+		USHORT PrintNameLength;
+		ULONG Flags;
+	} REPARSE_DATA_BUFFER, *PREPARSE_DATA_BUFFER_SYMLINK;
```
The definition is different from [here](https://docs.microsoft.com/en-us/windows-hardware/drivers/ddi/content/ntifs/ns-ntifs-_reparse_data_buffer).

After ULONG Flags, PathBuffer[1] is missing.

NoMoreFood replied:
This was left out purposefully but now that I think about it, adding it back in helps simplify the code.  See updated commit.

bagajjal commented on contrib/win32/win32compat/fileio.c:
```diff
@@ -749,13 +752,98 @@ fileio_fstat(struct w32_io* pio, struct _stat64 *buf)
+wchar_t * fileio_readlink_internal(wchar_t * wpath)
+{
+	/* abbreviated reparse data structure for symlinks */
+	typedef struct _REPARSE_DATA_BUFFER_SYMLINK {
```
This is not present in c++ SDK. It's interesting to know, how you got this information.

NoMoreFood replied:
It's primarily in the DDK: https://msdn.microsoft.com/en-us/library/cc232006.aspx

bagajjal commented on contrib/win32/win32compat/fileio.c:
```diff
@@ -749,13 +752,98 @@ fileio_fstat(struct w32_io* pio, struct _stat64 *buf)
+	}
+
+	/* trim off the extended drive prefix if present */
+	const wchar_t prefix[] = L"\\??\\";
```
[MSDN document](https://msdn.microsoft.com/en-us/library/cc422524.aspx) doesn't talk about the extended drive prefix. Also please add/explain the logic dealing with prefix[] with an example.

NoMoreFood replied:
See this: https://msdn.microsoft.com/en-us/library/windows/desktop/aa365247(v=vs.85).aspx.  I'll better characterize this in my code.

bagajjal replied:
Have you tried a symbolic link referencing a shared path? If yes, does it start with prefix "\\?\".

In code you used "\\?\\" instead of "\\?\", is it intentional? If yes, why?

NoMoreFood replied:
Good catch.  It comes back as `\??\UNC\` instead of `\??\C:\`.  I will have to address this as well.

NoMoreFood replied:
Addressed in latest commit.

bagajjal commented on contrib/win32/win32compat/fileio.c:
```diff
@@ -749,13 +752,98 @@ fileio_fstat(struct w32_io* pio, struct _stat64 *buf)
+
+	/* allocate an area for a null-terminated version of the string that can 
+	 * accommodate an absolute or relative path */
+	int linkpath_len = wcslen(wpath) + 1 + symlink_nonnull_size / sizeof(wchar_t) + 1;
```
This logic is bit confusing.
Using MAX_PATH improves readability.

bagajjal commented on contrib/win32/win32compat/fileio.c:
```diff
@@ -749,13 +752,98 @@ fileio_fstat(struct w32_io* pio, struct _stat64 *buf)
+		for (wchar_t * t = linkpath; *t != L'\0'; t++) if (*t == '/') *t = '\\';
+		wchar_t* path_end = linkpath + wcslen(linkpath);
+		while (*path_end != L'\\' && path_end > linkpath) (*path_end--) = L'\0';
+	}
```
as per BSD standards, else should start on the same line where if block curly brace ( } )ends .

NoMoreFood replied:
Reworked this code is latest commit.

bagajjal commented on contrib/win32/win32compat/fileio.c:
```diff
@@ -749,13 +752,98 @@ fileio_fstat(struct w32_io* pio, struct _stat64 *buf)
+	}
+
+	/* if string looks to be relative, copy the truncated parent */
+	if (symlink_nonnull[0] != L'\\' && !(symlink_nonnull_size / sizeof(wchar_t) >= 2 
```
Isn't this code sufficient? 
if (PathIsRelativeW(symlink_nonnull))
{
use _wfullpath() to get the absolute path
} else {
copy symlink_nonnull to linkpath
}

NoMoreFood replied:
The readlink() call doesn't necessarily limit to MAX_PATH so I'm trying avoid artificially shrinking it.  However, the reparse structure actually tells me if it's relative so I'm using that in the latest commit.

bagajjal commented on contrib/win32/win32compat/fileio.c:
```diff
@@ -749,13 +752,98 @@ fileio_fstat(struct w32_io* pio, struct _stat64 *buf)
+
+	/* if string looks to be relative, copy the truncated parent */
+	if (symlink_nonnull[0] != L'\\' && !(symlink_nonnull_size / sizeof(wchar_t) >= 2 
+		&& isalpha(symlink_nonnull[0]) && symlink_nonnull[1] == ':')) {
```
If (isalpha(symlink_nonnull[0]) && symlink_nonnull[1] == ':')) succeeds then it's an abosolute path.
But line 821 uses wpath, which looks wrong. Using wpath means it's treated as relative path.

NoMoreFood replied:
The 'wpath' that gets passed in always appears to be an absolute path.  Regardless, I reworked this code in the latest commit.

bagajjal commented on contrib/win32/win32compat/fileio.c:
```diff
@@ -944,4 +1053,60 @@ fileio_is_io_available(struct w32_io* pio, BOOL rd)
 	}
+}
+
+ssize_t fileio_readlink(const char *path, char *buf, size_t bufsiz)
```
please have return type, filename in different lines.

NoMoreFood replied:
Addressed in latest commit.

bagajjal commented on contrib/win32/win32compat/fileio.c:
```diff
@@ -944,4 +1053,60 @@ fileio_is_io_available(struct w32_io* pio, BOOL rd)
+
+	/* copy to output buffer - prepend forward slash */
+	convertToForwardslash(output);
+	buf[0] = '/';
```
setting buf[0] = '/', may not be required as we are sending windows absolute path.
The current code can handle windows paths.

NoMoreFood replied:
Based on other commits / issues, I had thought that the prepended /<DriveLetter>/ was decided as the standard.  Let me know if this isn't correct.

bagajjal replied:
ASAIK, standard doesn't say it.
I think its a good idea to have this code so that the unix code will treat it as absolute path.
Please ignore my comment.

bagajjal commented on contrib/win32/win32compat/misc.c:
```diff
@@ -831,13 +834,17 @@ w32_stat(const char *input_path, struct w32_stat *buf)
 	return fileio_stat(resolved_path(input_path), (struct _stat64*)buf);
 }
 
```
Since you already made changes to symbolic links, it would be great if you can implement w32_symlink(), link() implementation.

NoMoreFood replied:
I can probably look into those, but I'd like to get this pull in first before I work on those.

bagajjal commented:
Please add unit tests.

NoMoreFood committed:
Symbolic Link Code Cleanup

- Updated style to match upstream codebase.
- Added additional comments for clarity.
- Modified code to take advantage of relative flag within reparse symlink data.
- Simplified string parsing logic.

NoMoreFood commented:
Thank you for your review.  I still need to address the unit tests.

NoMoreFood committed:
Changed Symlink Target Path Lookup

- Modified fileio_readlink_internal() to use the symlink print name to avoid needing to accomodate different NT path formats for local versus network paths.
- Addressed small wording / formatting issues.

bagajjal commented on contrib/win32/win32compat/fileio.c:
```diff
@@ -749,12 +752,121 @@ fileio_fstat(struct w32_io* pio, struct _stat64 *buf)
+	 *
+	 * This approach uses the first method because the second method does not
+	 * work on broken link since the target file cannot be opened.  It also
+	 * requires additional I/O to read both the symlink and its target. */
```
Move the "*/" to new line.
Same comment for line 775

NoMoreFood replied:
Done in updated commit.

NoMoreFood replied:
Done in updated commit.

bagajjal commented on contrib/win32/win32compat/fileio.c:
```diff
@@ -749,12 +752,121 @@ fileio_fstat(struct w32_io* pio, struct _stat64 *buf)
+	/* allocate an area for a null-terminated version of the string; this can be
+	 * up to the length of the input path plus the relative path in the symlink */
+	const int wpath_len = wcslen(wpath);
+	int linkpath_len = wpath_len + 1 + symlink_nonnull_size / sizeof(wchar_t) + 1;
```
Using MAX_PATH makes this code simple and error free.

NoMoreFood replied:
So the code is a bit complicated for two reasons.  #1 The value returned by DeviceIoControl() is not null terminated so many standard functions will not process it.  This code avoids doing unnecessary copies of data by use the 'n' functions.  #2 The man page for readlink() notes that the contents of the symbolic link could be bigger than MAX_PATH so the buffer has to be allocated accordingly; the max size of the buffer is passed into readlink as bufsiz.

bagajjal commented on contrib/win32/win32compat/fileio.c:
```diff
@@ -749,12 +752,121 @@ fileio_fstat(struct w32_io* pio, struct _stat64 *buf)
 }
 
+wchar_t * 
+fileio_readlink_internal(wchar_t * wpath)
```
Does this logic supports both [Hard links](https://msdn.microsoft.com/en-us/library/windows/desktop/aa365006(v=vs.85).aspx) & soft links?

NoMoreFood replied:
No, readlink is only for symbolic links.

bagajjal commented on contrib/win32/win32compat/fileio.c:
```diff
@@ -749,12 +752,121 @@ fileio_fstat(struct w32_io* pio, struct _stat64 *buf)
+		/* copy the parent path, convert forward slashes to backslashes, and 
+		 * trim off the last entry in the path */
+		wcscpy_s(linkpath, linkpath_len, wpath);
+		for (int i = 0; i < wpath_len; i++)
```
Use convertToBackslashW().

NoMoreFood replied:
Done in updated commit.

NoMoreFood committed:
Symbolic Link Code Comment Style

- Changed comment style per review comments.
- Incorporated use of convertToBackslashW().
